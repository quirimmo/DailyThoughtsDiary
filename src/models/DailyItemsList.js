import moment from 'moment';
import DailyItems from './DailyItems';

class DailyItemsList extends Array {
	resetCounters = () => {
		this.forEach(dailyItem => dailyItem.resetCounter());
		return this;
	};

	addOrIncrementItem = item => {
		if (!this.containsDate(item.date)) {
			const dailyItems = new DailyItems(this.length.toString(), item.date);
			this.push(dailyItems);
		} else {
			this.getDailyItemsByDayDate(item.date).incrementItems();
		}
		return this;
	};

	containsDate = date => !!this.getDailyItemsByDayDate(date);

	getDailyItemsByDayDate = date =>
		this.find(el => moment(date).isSame(el.value, 'day'));
}

export default DailyItemsList;
