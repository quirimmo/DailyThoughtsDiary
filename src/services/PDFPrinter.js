import moment from 'moment';
import { Print } from 'expo';

class PDFPrinter {
	static async print(htmlContent) {
		try {
			await Print.printAsync({
				html: htmlContent
			});
		} catch (error) {
			console.error('Error printing with expo', error);
		}
	}

	static printDays(days) {
		const htmlContent = `<div><h1>All Days</h1>${getAllDays()}</div>`;
		PDFPrinter.print(htmlContent);

		function getAllDays() {
			return days
				.map(
					el =>
						`
						<div>
							<p>${moment(el.value).format('dddd DD-MM-YYYY')}: ${el.count} items</p>
						</div>
						`
				)
				.join('');
		}
	}

	static printItems(items) {
		const htmlContent = `<div><h1>All Items</h1>${getAllItems()}</div>`;
		PDFPrinter.print(htmlContent);

		function getAllItems() {
			return items
				.map(
					el =>
						`
						<div>
							<p>${el.symptoms}</p>
							<p>${el.where}</p>
							<p>${el.thoughts}</p>
							<p>${moment(el.value).format('dddd DD-MM-YYYY HH:mm')}</p>
							<br/>
						</div>
						`
				)
				.join('');
		}
	}
}

export default PDFPrinter;
