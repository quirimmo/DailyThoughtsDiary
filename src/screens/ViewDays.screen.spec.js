import React from 'react';
import { Alert } from 'react-native';
import NavigationTestUtils from 'react-navigation/NavigationTestUtils';
import renderer from 'react-test-renderer';
import DailyItemsList from '../models/DailyItemsList';

const items = [
	{
		date: '2018–10–28T12:34:56+00:00',
		symptoms: 'symptoms 1',
		location: 'location 1',
		thoughts: 'thoughts 1'
	},
	{
		date: '2018–10–29T12:34:56+00:00',
		symptoms: 'symptoms 2',
		location: 'location 2',
		thoughts: 'thoughts 2'
	}
];
const mockResetCounters = jest.fn(() => new DailyItemsList());
const mockAddOrIncrementItem = jest.fn(() => new DailyItemsList());
jest.mock('../models/DailyItemsList', () =>
	jest.fn().mockImplementation(() => ({
		resetCounters: mockResetCounters,
		addOrIncrementItem: mockAddOrIncrementItem
	}))
);

import ViewDaysScreen from './ViewDays.screen';
import PDFPrinter from '../services/PDFPrinter';

const mockFetchItems = jest.fn();
const mockNavigate = jest.fn();
const mockAddListener = jest.fn();
const navigation = {
	navigate: mockNavigate,
	addListener: mockAddListener
};
let component;

describe('ViewDaysScreen', () => {
	beforeEach(() => {
		jest.useFakeTimers();
		NavigationTestUtils.resetInternalState();
		component = renderer.create(
			<ViewDaysScreen
				navigation={navigation}
				fetchItems={mockFetchItems}
				items={items}
			/>
		);
	});
	afterEach(() => {
		jest.clearAllMocks();
	});

	it('should be defined', () => {
		expect(component).toBeDefined();
	});

	it('should render the component', async () => {
		expect(component.toJSON()).toMatchSnapshot();
	});

	describe('state', () => {
		describe('dailyItemsList', () => {
			it('should be defined', () => {
				expect(component.getInstance().state.dailyItemsList).toBeDefined();
			});
		});
	});

	describe('componentDidMount', () => {
		it('should call the addListener method of the navigation prop', () => {
			const instance = component.getInstance();
			instance.componentDidMount();
			expect(mockAddListener).toHaveBeenCalledWith(
				'didFocus',
				instance.loadItems
			);
		});
	});

	describe('loadItems success', () => {
		it('should call the resetCounters method of dailyItemsList', () => {
			component.getInstance().loadItems();
			expect(mockResetCounters).toHaveBeenCalled();
		});

		it('should call the fetchItems prop', () => {
			component.getInstance().loadItems();
			expect(mockFetchItems).toHaveBeenCalled();
		});

		it('should call the addOrIncrementItem method of DailyListItems', async () => {
			await component.getInstance().loadItems();
			expect(mockAddOrIncrementItem).toHaveBeenCalled();
		});
	});

	describe('loadItems error', () => {
		it('should display the error message', async () => {
			const spy = jest.spyOn(Alert, 'alert');
			const mockFetchItemsError = jest.fn(() => {
				throw new Error('fetch items error');
			});
			const fetchItemsErrorComponent = renderer.create(
				<ViewDaysScreen
					navigation={navigation}
					fetchItems={mockFetchItemsError}
					items={items}
				/>
			);
			await fetchItemsErrorComponent.getInstance().loadItems();
			expect(spy).toHaveBeenCalledWith(
				'Error fetching the items Error: fetch items error'
			);
		});
	});

	describe('printPDF', () => {
		it('should call the printDays method of PDFPrinter', () => {
			const spy = jest
				.spyOn(PDFPrinter, 'printDays')
				.mockImplementation(() => {});
			component.getInstance().printPDF();
			expect(spy).toHaveBeenCalledWith(
				component.getInstance().state.dailyItemsList
			);
		});
	});

	describe('openViewItems', () => {
		it('should call the navigate method of the navigation prop', () => {
			component.getInstance().openViewItems({ value: 'mocked-value' });
			expect(mockNavigate).toHaveBeenCalledWith('ViewItems', {
				date: 'mocked-value'
			});
		});
	});
});
