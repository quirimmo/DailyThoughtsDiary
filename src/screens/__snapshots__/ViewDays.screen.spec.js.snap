// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewDaysScreen should render the component 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <View>
    <View
      accessibilityComponentType="button"
      accessibilityLabel={undefined}
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={undefined}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Save All As PDF
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#e1e8ee",
          "height": 1,
        },
        Object {
          "marginVertical": 20,
        },
      ]
    }
  />
</View>
`;
